### PR TITLE
Program Settings: Custom browser command as command-line option (#806)

### PR DESCRIPTION
**Pull request by jolopezl**
Adds a command-line option to specify a browser and arguments for the browser. Follows #806.

The usage is simple:
```
./carta_backend --browser <browser>
```
where `<browser>` contains the entire system call to open the front end URL. The user can use the placeholder `%url` to add a custom position for the URL, for example, before or after browser arguments. If the `%url` wildcard is not given, the URL default position is at the end. For instance, a general case looks like this:
```
./carta_backend --browser "firefox -CreateProfile CARTAuser %url"
```
The `browser` option can also be passed using system and user JSON configuration files. As with the other command-line settings, just add:
```
{
    "browser" : "open -a firefox"
}
```
Note that for macOS, Google Chrome needs to be spelled `Google\ Chrome`, which means `Google\\ Chrome` in the JSON file; otherwise, the parsing will fail.

**Working/tested**: The following are examples tested and working on macOS:
```
./carta_backend --verbosity 4 --frontend_folder ~/temp/CARTAvis/carta-frontend/build \
 --root ~/temp/CARTAvis/data --base ~/temp/CARTAvis/data --omp_threads 4 \
 --browser "open -a safari"
```
```
./carta_backend --verbosity 4 --frontend_folder ~/temp/CARTAvis/carta-frontend/build \
 --root ~/temp/CARTAvis/data --base ~/temp/CARTAvis/data --omp_threads 4 \
 --browser "open -a firefox"
```
```
./carta_backend --verbosity 4 --frontend_folder ~/temp/CARTAvis/carta-frontend/build \
 --root ~/temp/CARTAvis/data --base ~/temp/CARTAvis/data --omp_threads 4 \
 --browser "open -a Google\ Chrome"
```
**TO DO:**
- [ ] Test functionality on linux.
- [ ] Need a test?

**Open question:**
- [ ] The issued command is the full responsibility of the user. Do we want some minimal checking?

### CODE REVIEW

**Comment by kswang1029** [comment]
@jolopezl can the following be simplified from
```
--browser "open -a safari"
--browser "open -a firefox"
--browser "open -a Google\ Chrome"
```
to
```
--browser safari
--browser firefox
--browser chrome
```
?

this might be more friendly to users?

**Comment by ajm-ska** [comment]
It looks like if I have Chrome already open, and Chrome is set as the default browser, then doing `./carta_backend --browser "open -a safari`, opens CARTA on both Chrome and Safari.
If Safari is set as the default browser (and Chrome is already open), then CARTA only opens in Safari with the above command. 
Although if Chrome is the default browser and already open, then doing `./carta_backend --browser "open -a firefox` only opens CARTA in Firefox.

**Comment by jolopezl** [comment]
> @jolopezl can the following be simplified from
> 
> ```
> --browser "open -a safari"
> --browser "open -a firefox"
> --browser "open -a Google\ Chrome"
> ```
> 
> to
> 
> ```
> --browser safari
> --browser firefox
> --browser chrome
> ```
> 
> ?
> 
> this might be more friendly to users?

We discussed this briefly and thought it's the user's responsibility to give the correct browser command and its arguments. The appropriate command depends on the system and can change with different versions, etc.

**Comment by kswang1029** [comment]
> > @jolopezl can the following be simplified from
> > ```
> > --browser "open -a safari"
> > --browser "open -a firefox"
> > --browser "open -a Google\ Chrome"
> > ```
> > 
> > 
> > to
> > ```
> > --browser safari
> > --browser firefox
> > --browser chrome
> > ```
> > 
> > 
> > ?
> > this might be more friendly to users?
> 
> We discussed this briefly and thought that it's the user's responsibility to give the right browser command and its arguments. The appropriate command depends on the system and can change with different versions, etc.

ok. I was thinking if we can make the backend smarter on this aspect. The backend should be able to tell what OS it is then trigger the corresponding command under the scene.

**Comment by jolopezl** [comment]
> > > @jolopezl can the following be simplified from
> > > ```
> > > --browser "open -a safari"
> > > --browser "open -a firefox"
> > > --browser "open -a Google\ Chrome"
> > > ```
> > > 
> > > 
> > > to
> > > ```
> > > --browser safari
> > > --browser firefox
> > > --browser chrome
> > > ```
> > > 
> > > 
> > > ?
> > > this might be more friendly to users?
> > 
> > 
> > We discussed this briefly and thought that it's the user's responsibility to give the right browser command and its arguments. The appropriate command depends on the system and can change with different versions, etc.
> 
> ok. I was thinking if we can make the backend smarter on this aspect. The backend should be able to tell what OS it is then trigger the corresponding command under the scene.

Of course, we can, but it could break anytime in the future between releases if firefox decides to start calling himself firefox-browser. We can discuss adding support now for Ubuntu 18.04 and 20.04, Centos 7 and 8, and macOS, but I foresee hotfixes to patch for system or browser changes.

**Comment by veggiesaurus** [comment]
> @jolopezl can the following be simplified from
> 
> ```
> --browser "open -a safari"
> --browser "open -a firefox"
> --browser "open -a Google\ Chrome"
> ```
> 
> to
> 
> ```
> --browser safari
> --browser firefox
> --browser chrome
> ```
> 
> ?
> 
> this might be more friendly to users?

I'd rather leave this up to the users. We give them the option to provide whatever command they like, and it's up to them to get that command right. Jupyter notebook takes a similar approach to this. There are just too many variables and headaches that could occur if we try to be smart about it. They might have an alias to the applications in their bash profile, for example. Or have Chromium instead of Google Chrome, or have Safari Tech Preview, etc etc.

**Comment by jolopezl** [comment]
> It looks like if I have Chrome already open, and Chrome is set as the default browser, then doing `./carta_backend --browser "open -a safari`, opens CARTA on both Chrome and Safari.
> If Safari is set as the default browser (and Chrome is already open), then CARTA only opens in Safari with the above command.
> Although if Chrome is the default browser and already open, then doing `./carta_backend --browser "open -a firefox` only opens CARTA in Firefox.

Thanks for looking at it. Although, I can't reproduce the first case:

> It looks like if I have Chrome already open, and Chrome is set as the default browser, then doing `./carta_backend --browser "open -a safari`, opens CARTA on both Chrome and Safari.

Could you please, confirm? I tried:
- setting Google Chrome as the default browser,
- tested default browser by doing: `open https://www.gmail.com`; it opens Gmail on Chrome,
- running `./carta_backend --browser "open -a safari`, and it pops up Safari, without opening Chrome.

File: src/Main.cc
Comment (veggiesaurus) [inline src/Main.cc]: I don't think we need this output, or at least we can leave it as debug

File: src/SessionManager/ProgramSettings.h
Comment (veggiesaurus) [inline src/SessionManager/ProgramSettings.h]: I think `GenerateBrowserCommand` makes more sense here. It's parsing the browser _argument_ to generate a command string

**Review by veggiesaurus — CHANGES_REQUESTED** [review]

**Comment by veggiesaurus** [comment]
> > It looks like if I have Chrome already open, and Chrome is set as the default browser, then doing `./carta_backend --browser "open -a safari`, opens CARTA on both Chrome and Safari.
> > If Safari is set as the default browser (and Chrome is already open), then CARTA only opens in Safari with the above command.
> > Although if Chrome is the default browser and already open, then doing `./carta_backend --browser "open -a firefox` only opens CARTA in Firefox.
> 
> Thanks for looking at it. Although, I can't reproduce the first case:
> 
> > It looks like if I have Chrome already open, and Chrome is set as the default browser, then doing `./carta_backend --browser "open -a safari`, opens CARTA on both Chrome and Safari.
> 
> Could you please, confirm? I tried:
> 
> * setting Google Chrome as the default browser,
> * tested default browser by doing: `open https://www.gmail.com`; it opens Gmail on Chrome,
> * running `./carta_backend --browser "open -a safari`, and it pops up Safari, without opening Chrome.

I can't reproduce this either. Since this is up to the user to get right, something like `--browser="/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome"` would call the application directly, instead of using `open`

**Comment by ajm-ska** [comment]
Here is a video of it happening to me:
https://user-images.githubusercontent.com/20802551/117416976-7cbbf480-af4c-11eb-9c2f-0f291e2bb63f.mov

File: src/SessionManager/ProgramSettings.cc
Comment (veggiesaurus) [inline src/SessionManager/ProgramSettings.cc]: shouldn't it be `"use custom browser + args"`? I prefer something along the lines of 

```
Specify command to use to invoke a web browser when opening the CARTA frontend. If not specified, the default browser is used.
```

Comment (confluence) [reply src/SessionManager/ProgramSettings.cc]: We should not be putting long explanations in the command descriptions. Any long explanation should go in the body of the help.

Comment (confluence) [inline src/SessionManager/ProgramSettings.cc]: I thought the intention was to merge these into a single option, and use `--browser=none`, but maybe it's fine to leave this as two options. On the one hand, I would prefer not to have a proliferation of related and overlapping options. On the other hand, checking explicitly for a "none" keyword is a bit hacky.

We should add in the body of the help that `no_browser` takes precedence over `browser`.

Comment (veggiesaurus) [reply src/SessionManager/ProgramSettings.cc]: fair enough. But `user -> use` makes sense, right?

Comment (kswang1029) [reply src/SessionManager/ProgramSettings.cc]: can we have _one_ example for args so that users can be inspired for the alternative options?

File: src/SessionManager/ProgramSettings.h
Comment (confluence) [inline src/SessionManager/ProgramSettings.h]: We can't use `%url`; this will be interpreted by FreeDesktop as `%u` in a desktop file on Linux. I would prefer something simpler like `CARTA_URL`.

File: src/SessionManager/ProgramSettings.cc
Comment (confluence) [reply src/SessionManager/ProgramSettings.cc]: I suggest  `custom browser command` here, with an explanation later; something like:

`Disabling the browser takes precedence over a custom browser command. The custom browser command may contain the placeholder CARTA_URL, which will be replaced by the frontend URL. If the placeholder is omitted, the URL will be appended to the end.`

(Or whatever we use for the placeholder, which can't be `%url`, as I explained in the other comment.)

Edit: I prefer `custom browser command` to `use custom browser command` because the latter makes it sound like a boolean.

**Review by confluence — CHANGES_REQUESTED** [review]

File: src/SessionManager/ProgramSettings.cc
Comment (veggiesaurus) [reply src/SessionManager/ProgramSettings.cc]: initially I'd wanted that too but @jolopezl convinced me otherwise :smile:

FWIW jupyter also uses two separate args

**Comment by confluence** [comment]
General comment: I think that we should absolutely leave this to users and make no attempt to check for correctness, which is highly environment-specific. We can build on this to create custom launchers (and suggest specific recipes for using the option in the documentation) to help users transition away from the bundled electron app, but the option should not be restricted to those use cases.

File: src/SessionManager/ProgramSettings.cc
Comment (confluence) [reply src/SessionManager/ProgramSettings.cc]: Makes sense.

Comment (confluence) [reply src/SessionManager/ProgramSettings.cc]: > can we have _one_ example for args so that users can be inspired for the alternative options?

I would really prefer not to put that in the executable help. It's much too specific for this context. I do think that we should add some information about this to the _manual_, and we can give lots of examples there.

File: src/SessionManager/ProgramSettings.h
Comment (jolopezl) [reply src/SessionManager/ProgramSettings.h]: Fixed.

File: src/SessionManager/ProgramSettings.h
Comment (jolopezl) [reply src/SessionManager/ProgramSettings.h]: Changed.

**Comment by confluence** [comment]
This looks good; I'm just building this so that I can test the Linux desktop file use case.

**Comment by jolopezl** [comment]
> Here is a video of it happening to me:
> https://user-images.githubusercontent.com/20802551/117416976-7cbbf480-af4c-11eb-9c2f-0f291e2bb63f.mov

I still can't reproduce. I also tested with zsh. What terminal are you using? Maybe some underlying instruction to open links as they appear?

**Comment by confluence** [comment]
I'm seeing some weird behaviour trying to use Chrome (actually Chromium) on Linux.

If I run the backend with `--no-browser`, and then manually run `chromium-browser --new-window --app=URL_WITH_TOKEN_GOES_HERE`, I get the expected result.

However, if I run the backend with `--browser="chromium-browser --new-window --app=CARTA_URL"`, I get a blank window and CARTA never loads. I also don't see `CARTA is accessible at...` in the backend output until after I close the Chromium window.

I tested this with `firefox --new-window` as the browser, and that works fine, but chromium doesn't work at all, even with no parameters. However, if I explicitly run the chromium process in the background, by making the browser command `--browser="chromium-browser --app=CARTA_URL --new-window&"`, it works as expected (and I have confirmed that I can make a custom launcher with a desktop file using this command).

So it looks like the chromium process is blocking the backend process from continuing. But why (when Firefox doesn't behave like this), is there anything we can do about this, and is it our problem? Is explicitly launching it in the background an acceptable workaround?

@veggiesaurus and anyone else with a Linux desktop, can you reproduce this, or is it just me?

**Comment by confluence** [comment]
OK, I think this is a non-mystery mystery -- I checked again and when I compare launching browsers on the commandline, Chromium stays in the foreground and Firefox does not. So I think this is not our problem, and we just need to remember to background the process explicitly in the command.

**Comment by confluence** [comment]
@veggiesaurus has pointed out that it may be a good idea for us _always_ to fork the browser process.

**Comment by confluence** [comment]
> @veggiesaurus and anyone else with a Linux desktop, can you reproduce this, or is it just me?

Apparently this doesn't happen with Chrome, so it may be behaviour specific to Chromium.

**Comment by veggiesaurus** [comment]
> > @veggiesaurus and anyone else with a Linux desktop, can you reproduce this, or is it just me?
> 
> Apparently this doesn't happen with Chrome, so it may be behaviour specific to Chromium.

wait, no. It _does_ happen with Chrome. It happens with `google-chrome`, `chromium` and `firefox` for me.

I wonder if the easiest solution is to just append a ` &` to the command string _if_ it doesn't already end with it (and to not change anything if there's no `--browser` argument.

**Comment by confluence** [comment]
> I wonder if the easiest solution is to just append a ` &` to the command string

How portable is that? And should we also suppress the output?

**Comment by jolopezl** [comment]
@confluence, @veggiesaurus, I just checked on Ubuntu 18.04, and 20.04 (VMs), and the wrong behavior happens with Chrome and Chromium, but it doesn't happen with Firefox. But, as @confluence suggested, if I add `&` at the end of the command, Chrome or Chromium works perfectly.

So, we see all kinds of mixed behaviors. May I suggest that we leave this low-level setting to the end-user? We should describe this subtle behavior in the documentation rather than trying to get it right for all.

We can force the process to the background, of course. Do we anticipate side effects? As @confluence says, we would need to redirect the output as well.

What's your preference? I prefer to let the user get the command right.

**Comment by confluence** [comment]
I think that we should ensure that the command is always run as an independent process which will remain alive if the terminal in which the backend is running is closed -- because this is consistent with the behaviour of `open` / `xdg-open`, and because if we don't do this and the process stays in the foreground it causes very counter-intuitive and user-unfriendly behaviour.

But I would prefer to do this in a cleaner C++ way than to tack on a `&` to the shell command. Just the `&` is not sufficient for the correct behaviour -- that will not disown the process, so it will be closed when the terminal is closed (there are various ways to make that not happen, e.g. the `disown` shell command). From a cursory look at Jupyter notebook (which uses `webbrowser`, which uses `subprocess`) we could use `setsid` to run the process independently. I think we should also put some thought into what happens to the output.

I think @veggiesaurus is currently suggesting that we merge this as-is and refine it during the beta in time for the final 2.0 release -- I would prefer to do this than to rush in a hacky fix. In the meantime, we can document this limitation in the information we give to beta testers, and suggest temporary shell workarounds.

**Comment by jolopezl** [comment]
> I think @veggiesaurus is currently suggesting that we merge this as-is and refine it during the beta in time for the final 2.0 release -- I would prefer to do this than to rush in a hacky fix. In the meantime, we can document this limitation in the information we give to beta testers, and suggest temporary shell workarounds.

Ok, I agree. Let's merge this – if you all approve – and provide a complete solution for the final 2.0 release.

**Review by confluence — APPROVED** [review]

**Comment by ajm-ska** [comment]
> > Here is a video of it happening to me:
> > https://user-images.githubusercontent.com/20802551/117416976-7cbbf480-af4c-11eb-9c2f-0f291e2bb63f.mov
> 
> I still can't reproduce. I also tested with zsh. What terminal are you using? Maybe some underlying instruction to open links as they appear?

I'm using zsh. I'm not sure if the version makes a difference, but it is Chrome 90.0.4430.93.

If I try to open Safari directly with ` --browser "/Applications/Safari.app/Contents/MacOS/Safari"` then Chrome doesn't open, the Safari window opens, but it hangs because there is an error in the terminal:

> objc[35741]: Class AMSupportURLConnectionDelegate is implemented in both /usr/lib/libauthinstall.dylib (0x7fff83050ac0) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x11158a910). One of the two will be used. Which one is undefined.
> objc[35741]: Class AMSupportURLSession is implemented in both /usr/lib/libauthinstall.dylib (0x7fff83050b10) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x11158a960). One of the two will be used. Which one is undefined.

Does that work for everyone else?

Another thing I notice on MacOS is if Firefox is closed and I use  `--browser "open -a firefox"` then two Firefox windows are actually opened; A blank window and a window with CARTA.

Screenshot 2021-05-10 at 08 34 24

I can also report things work fine with Edge. Doing ` --browser "open -a Microsoft\ Edge" ` only opens CARTA in Edge.

**Review by veggiesaurus — APPROVED** [review]

**Comment by veggiesaurus** [comment]
I'm going to merge this in before code freeze. I suggest we make an issue for improving the behaviour of the `--browser` flag, and see if we can fix it before the 2.0 release itself.

**Comment by jolopezl** [comment]
@ajm-asiaa, as a minor follow up to your finding:

> Another thing I notice on MacOS is if Firefox is closed and I use `--browser "open -a firefox"` then two Firefox windows are actually opened; A blank window and a window with CARTA.

This seems to be an ancient bug: https://bugzilla.mozilla.org/show_bug.cgi?id=531552, which is 12 years old with some recent reports of the same happening, and new bugs filed: https://bugzilla.mozilla.org/show_bug.cgi?id=1627403, https://bugzilla.mozilla.org/show_bug.cgi?id=1600153. It seems to be something different, but the effect is the same. I am reporting this so that we don't worry too much about it for now.

**Pull request merged**
